fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/subsumes/ClosureService.java

```diff
@@ -0,0 +1,123 @@
+package au.csiro.pathling.fhirpath.function.subsumes;
+
+import au.csiro.pathling.fhir.TerminologyClient;
+import au.csiro.pathling.fhirpath.encoding.SimpleCoding;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import lombok.extern.slf4j.Slf4j;
+import org.hl7.fhir.r4.model.Coding;
+import org.hl7.fhir.r4.model.ConceptMap;
+import org.hl7.fhir.r4.model.ConceptMap.ConceptMapGroupComponent;
+import org.hl7.fhir.r4.model.ConceptMap.SourceElementComponent;
+import org.hl7.fhir.r4.model.ConceptMap.TargetElementComponent;
+import org.hl7.fhir.r4.model.Enumerations.ConceptMapEquivalence;
+import org.hl7.fhir.r4.model.Enumerations.PublicationStatus;
+import org.hl7.fhir.r4.model.StringType;
+
+/**
+ * Helper class to encapsulate the creation of a subsumes relation for a set of Codings.
+ *
+ * @author Piotr Szul
+ */
+@Slf4j
+class ClosureService {
+
+  @Nonnull
+  private final TerminologyClient terminologyClient;
+
+  public ClosureService(@Nonnull TerminologyClient terminologyClient) {
+    this.terminologyClient = terminologyClient;
+  }
+
+  @Nonnull
+  public Closure getSubsumesRelation(
+      Collection<SimpleCoding> systemAndCodes) {
+    List<Coding> codings =
+        systemAndCodes.stream().map(SimpleCoding::toCoding).collect(Collectors.toList());
+    // recreate the systemAndCodes dataset from the list not to execute the query again.
+    // Create a unique name for the closure table for this code system, based upon the
+    // expressions of the input, argument and the CodeSystem URI.
+    String closureName = UUID.randomUUID().toString();
+    log.info("Sending $closure request to terminology service with name '{}' and {} codings",
+        closureName, codings.size());
+    // Execute the closure operation against the terminology server.
+    // TODO: add validation checks for the response
+    ConceptMap initialResponse = terminologyClient.closure(new StringType(closureName), null);
+    validateConceptMap(initialResponse, closureName, "1");
+    ConceptMap closureResponse =
+        terminologyClient.closure(new StringType(closureName), codings);
+    validateConceptMap(closureResponse, closureName, "2");
+    return conceptMapToClosure(closureResponse);
+  }
+
+  private void validateConceptMap(ConceptMap conceptMap, String closureName, String version) {
+    if (!PublicationStatus.ACTIVE.equals(conceptMap.getStatus())) {
+      throw new RuntimeException(
+          "Expected ConceptMap with status: ACTIVE, got: " + conceptMap.getStatus());
+    }
```

Comment:
I don't think this validation is necessary.

I think it is very possible (and valid) that users will want to use ConceptMap resources with other statuses (particularly `draft`) with Pathling. I don't think we should be opinionated on the value of the `status` element.

The type of validation we may need is checking that the content that we rely upon within the ConceptMap is complete, e.g. check for missing `element` in `group`, check for missing `target` in `element`, etc.

Reply:
OK.